### PR TITLE
[WIP] Add layer filter accessor

### LINKED ISSUE #7978: Add filter accessor to layer
ATM there is only a setter for filters on `Layer` classes, no getter, eg `Layer#getFilter()`.

### PR DESCRIPTION
**Pull request by Guardiola31337**
Fixes #7978 

cc/ @ivovandongen @tobrun

### CODE REVIEW

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/layers/layer.java.ejs [hunk 1]
```diff
@@ -89,6 +89,15 @@ public class <%- camelize(type) %>Layer extends Layer {
   }
 
   /**
+   * Gets the filter of the layer.
+   *
+   * @return {@link Filter.Statement} filter layer is using
+   */
+  public Filter.Statement getFilter() {
```
Comment (ivovandongen) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/layers/layer.java.ejs]: `@Nullable`?

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/layers/layer.java.ejs [hunk 2]
```diff
@@ -89,6 +89,15 @@ public class <%- camelize(type) %>Layer extends Layer {
   }
 
   /**
+   * Gets the filter of the layer.
+   *
+   * @return {@link Filter.Statement} filter layer is using
```
Comment (ivovandongen) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/layers/layer.java.ejs]: "or null"

File: platform/android/src/style/layers/layer.cpp [hunk 3]
```diff
@@ -111,14 +112,31 @@ namespace android {
         }
     };
 
+    jni::Object<mbgl::android::Statement> Layer::getFilter(jni::JNIEnv& env) {
+        using namespace mbgl::style;
+        using namespace mbgl::android::conversion;
+
+        if (layer.is<FillLayer>()) {
```
Comment (ivovandongen) [inline platform/android/src/style/layers/layer.cpp]: Please replace with visitor like in: https://github.com/mapbox/mapbox-gl-native/pull/9069

File: platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/testapp/style/FillLayerTest.java [hunk 4]
```diff
@@ -65,6 +67,17 @@ private void setupLayer(){
   }
 
   @Test
+  public void testGetFilter() {
```
Comment (ivovandongen) [inline platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/testapp/style/FillLayerTest.java]: This file is generated. Filter tests should either be generated too or at least live in another file so they are not overwritten.

File: platform/android/src/style/conversion/filter.hpp [hunk 5]
```diff
@@ -27,6 +30,16 @@ inline optional<mbgl::style::Filter> toFilter(jni::JNIEnv& env, jni::Array<jni::
     return filter;
 }
 
+template <>
+struct Converter<jni::Object<mbgl::android::Statement>, mbgl::style::Filter> {
+    Result<jni::Object<mbgl::android::Statement>> operator()(jni::JNIEnv& env, const mbgl::style::Filter& value) const {
+        using namespace mbgl::android;
+        auto statement = Filter::fromFilter(env, value);
```
Comment (ivovandongen) [inline platform/android/src/style/conversion/filter.hpp]: Don't need a temporary variable here

File: platform/android/src/style/filter.cpp [hunk 6]
```diff
@@ -0,0 +1,21 @@
+#include "filter.hpp"
+
+namespace mbgl {
+namespace android {
+
+jni::Object<Statement> Filter::fromFilter(jni::JNIEnv& env, const mbgl::style::Filter&) {
+    static auto constructor = Statement::javaClass.template GetConstructor<>(env);
+
```
Comment (ivovandongen) [inline platform/android/src/style/filter.cpp]: nit : blank line

File: platform/android/src/style/statement.cpp [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#include "statement.hpp"
+
+namespace mbgl {
+namespace android {
+
+void Statement::registerNative(jni::JNIEnv& env) {
```
Comment (ivovandongen) [inline platform/android/src/style/statement.cpp]: This is not actually called

File: platform/android/src/style/statement.hpp [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+#include <mbgl/style/filter.hpp>
+#include <mbgl/util/noncopyable.hpp>
+
+#include <jni/jni.hpp>
+
+namespace mbgl {
+namespace android {
+
+class Statement : private mbgl::util::noncopyable {
```
Comment (ivovandongen) [inline platform/android/src/style/statement.hpp]: Please wrap this in `Filter`, so it mirrors the Java class hierarchy and reads better

**Comment by tobrun** [comment]
Moving forward on this, functions are going to be deprecated with the introduction with expressions. Closing as obsolete.

**Pull request closed without merging**